internal/utils/utils_test.go

```diff
@@ -172,3 +178,81 @@ var _ = Describe("InheritNotReadyConditionFromObjects", func() {
 		Expect(readyCondition.Message).To(Equal(`CommitStatus "child1" is not Ready because "NotReady1": Child1 is not ready`))
 	})
 })
+
+var _ = Describe("HandleReconciliationResult panic recovery", func() {
+	var (
+		ctx      context.Context
+		obj      *promoterv1alpha1.PromotionStrategy
+		recorder record.EventRecorder
+		scheme   *runtime.Scheme
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+		obj = &promoterv1alpha1.PromotionStrategy{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "PromotionStrategy",
+				APIVersion: "promoter.argoproj.io/v1alpha1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:       "test-strategy",
+				Namespace:  "default",
+				Generation: 1,
+			},
+		}
+		scheme = runtime.NewScheme()
+		_ = promoterv1alpha1.AddToScheme(scheme)
+		recorder = record.NewFakeRecorder(10)
+	})
+
+	It("should recover from panic and convert it to an error", func() {
+		var err error
+		fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithStatusSubresource(obj).Build()
```

Comment:
```suggestion
		// We use fakeclient here since it's virtually impossible to trigger a panic otherwise. Don't spread
		// this use to other tests if at all avoidable.
		fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithStatusSubresource(obj).Build()
```